packages/core/src/actors/promise.ts

```diff
@@ -132,29 +139,38 @@ export function fromPromise<TOutput, TInput = NonReducibleUnknown>(
       }
 
       const resolvedPromise = Promise.resolve(
-        promiseCreator({ input: state.input!, system, self })
+        promiseCreator({
+          input: state.input!,
+          system,
+          self,
+          signal: controller.signal
+        })
       );
 
-      resolvedPromise.then(
-        (response) => {
-          if (self.getSnapshot().status !== 'active') {
-            return;
-          }
-          system._relay(self, self, {
-            type: XSTATE_PROMISE_RESOLVE,
-            data: response
-          });
-        },
-        (errorData) => {
-          if (self.getSnapshot().status !== 'active') {
-            return;
+      resolvedPromise
+        .then(
+          (response) => {
+            if (self.getSnapshot().status !== 'active') {
+              return;
+            }
+            system._relay(self, self, {
+              type: XSTATE_PROMISE_RESOLVE,
+              data: response
+            });
+          },
+          (errorData) => {
+            if (self.getSnapshot().status !== 'active') {
+              return;
+            }
+            system._relay(self, self, {
+              type: XSTATE_PROMISE_REJECT,
+              data: errorData
+            });
           }
-          system._relay(self, self, {
-            type: XSTATE_PROMISE_REJECT,
-            data: errorData
-          });
-        }
-      );
+        )
+        .finally(() => {
+          done = true;
+        });
```

Comment:
To avoid potential race conditions between scheduled microtasks, I would strongly prefer for this to be set asap - within each then/catch callback. Notice how this microtask is executed *before* the finally handler:
```ts
Promise.resolve(42)
  .then(() => {
    queueMicrotask(() => console.log("micro"));
  })
  .finally(() => console.log("end"));
```

Reply:
And actually, maybe you could just clear `controller` instead of introducing a new `done` variable.

Reply:
[`c06b029` (#4832)](https://github.com/statelyai/xstate/pull/4832/commits/c06b029019d1aac6d9d66166a5a940c4f49661f8)